### PR TITLE
Report collector command to add DP noise

### PR DESCRIPTION
**Pull request by akoshelev**
This adds a new command to the report collector binary that takes the output file created by IPA commands and applies noise according to (epsilon, delta, cap) values provided as arguments to this command.

After discussing this with @bmcase I got convinced that we need unsigned integers to represent values with added noise because otherwise we may introduce bias to the results (moving mean to the right) that may prevent further aggregations. 

## Sample command

```bash
 ./target/release/report_collector --input-file /tmp/breakdowns.log --output-file /tmp/noise.log apply-dp-noise -e 1.0 -e 2.0 -e 3.0 -c 5
```

## Sample output

```bash
+---------+----------+-------+-----+-----+-----+-----+-----+-----+-----+----+-----+-----+-----+----+-----+-----+----+-----+-----+----+-----+-----+
| Epsilon | Variance | Mean  | 1   | 2   | 3   | 4   | 5   | 6   | 7   | 8  | 9   | 10  | 11  | 12 | 13  | 14  | 15 | 16  | 17  | 18 | 19  | 20  |
+================================================================================================================================================+
| -       | -        | -     | 103 | 120 | 119 | 151 | 84  | 0   | 0   | 0  | 0   | 0   | 0   | 0  | 0   | 0   | 0  | 0   | 0   | 0  | 0   | 0   |
|---------+----------+-------+-----+-----+-----+-----+-----+-----+-----+----+-----+-----+-----+----+-----+-----+----+-----+-----+----+-----+-----|
| 3.000   | 9.528    | 0.000 | 110 | 123 | 113 | 145 | 108 | -20 | -11 | -8 | -6  | -6  | -3  | 14 | 9   | 5   | -3 | -10 | 6   | 14 | -4  | 4   |
|---------+----------+-------+-----+-----+-----+-----+-----+-----+-----+----+-----+-----+-----+----+-----+-----+----+-----+-----+----+-----+-----|
| 2.000   | 14.292   | 0.000 | 108 | 117 | 118 | 168 | 80  | -1  | 8   | 14 | -19 | -14 | 1   | 12 | -15 | -3  | 19 | 17  | -18 | 8  | -4  | 12  |
|---------+----------+-------+-----+-----+-----+-----+-----+-----+-----+----+-----+-----+-----+----+-----+-----+----+-----+-----+----+-----+-----|
| 1.000   | 28.584   | 0.000 | 89  | 65  | 155 | 150 | 87  | -1  | -18 | -9 | 17  | -18 | -31 | -7 | -18 | -13 | 30 | 20  | -42 | 28 | -19 | -20 |
+---------+----------+-------+-----+-----+-----+-----+-----+-----+-----+----+-----+-----+-----+----+-----+-----+----+-----+-----+----+-----+-----+
```

json file

```json
{
  "1": {
    "breakdowns": [
      89,
      65,
      155,
      150,
      87,
      -1,
      -18,
      -9,
      17,
      -18,
      -31,
      -7,
      -18,
      -13,
      30,
      20,
      -42,
      28,
      -19,
      -20
    ],
    "mean": 0.0,
    "std": 28.584295690354633
  },
  "2": {
    "breakdowns": [
      108,
      117,
      118,
      168,
      80,
      -1,
      8,
      14,
      -19,
      -14,
      1,
      12,
      -15,
      -3,
      19,
      17,
      -18,
      8,
      -4,
      12
    ],
    "mean": 0.0,
    "std": 14.292147845177317
  },
  "3": {
    "breakdowns": [
      110,
      123,
      113,
      145,
      108,
      -20,
      -11,
      -8,
      -6,
      -6,
      -3,
      14,
      9,
      5,
      -3,
      -10,
      6,
      14,
      -4,
      4
    ],
    "mean": 0.0,
    "std": 9.528098563451545
  }
}
```

### CODE REVIEW

**Comment by bmcase** [comment]
>  After discussing this with @bmcase I got convinced that we need unsigned integers to represent values with added noise because otherwise we may introduce bias to the results (moving mean to the right) that may prevent further aggregations.

I think you mean we need signed integers

**Review by benjaminsavage — APPROVED** [review]

File: src/cli/noise.rs [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+pub struct NoisyOutput {
+    /// Aggregated breakdowns with noise applied. It is important to use unsigned values here
+    /// to avoid bias/mean skew
+    pub breakdowns: Box<[i64]>,
+    pub mean: f64,
+    pub std: f64,
+}
+
+/// This exists to be able to use f64 as key inside a map. We don't have to deal with infinities or
+/// NaN values for epsilons so
```
Comment (benjaminsavage) [inline src/cli/noise.rs:42]: so... what?

**Review by richajaindce — APPROVED** [review]
Looks correct directionally, lets land it since it is needed for the test

**Comment by benjaminsavage** [comment]
revealing the standard deviation doesn't leak additional info right?

**Comment by akoshelev** [comment]
> revealing the standard deviation doesn't leak additional info right?
yea we reveal aggregates in the clear and we are not supposed to run this DP in prod. This command exists to illustrate the effects of adding noise to real data

**Comment by bmcase** [comment]
> revealing the standard deviation doesn't leak additional info right?
No, you can get this directly from epsilon, delta, and sensitivity.  It's just shown for convenience in interpreting the results when playing with different epsilons and caps.

**Pull request merged**